### PR TITLE
Timeouts

### PR DESCRIPTION
**Pull request by ubuntolog**
Added timeouts for container creation and execution (including swarm services). Solves the issue https://github.com/EUDAT-GEF/GEF/issues/92

### CODE REVIEW

File: gefserver/pier/internal/dckr/dckr.go [hunk 1]
```diff
@@ -368,13 +370,18 @@ func (c Client) StartImage(id string, repoTag string, cmdArgs []string, binds []
 		MemorySwap: limits.MemorySwap,
 	}
 
+	createContainerContext, cancel := context.WithTimeout(context.Background(), time.Duration(preparationTimeOut)*time.Second)
+	defer cancel()
```
Comment (emanueldima) [inline gefserver/pier/internal/dckr/dckr.go]: What does this do?

File: gefserver/pier/internal/dckr/dckr.go [hunk 2]
```diff
@@ -368,13 +370,18 @@ func (c Client) StartImage(id string, repoTag string, cmdArgs []string, binds []
...
+
 	cco := docker.CreateContainerOptions{
 		Config:     &config,
 		HostConfig: &hc,
+		Context:    createContainerContext,
 	}
 
 	cont, err := c.c.CreateContainer(cco)
 	if err != nil {
+		log.Println(err)
```
Comment (emanueldima) [inline gefserver/pier/internal/dckr/dckr.go]: ?

File: gefserver/pier/internal/dckr/dckr.go [hunk 3]
```diff
@@ -395,7 +402,9 @@ func (c Client) StartImage(id string, repoTag string, cmdArgs []string, binds []
 	<-attached
 	attached <- struct{}{}
 
-	err = c.c.StartContainer(cont.ID, &hc)
+	jobExecutionContext, cancel := context.WithTimeout(context.Background(), time.Duration(executionTimeOut)*time.Second)
+	defer cancel()
```
Comment (emanueldima) [inline gefserver/pier/internal/dckr/dckr.go]: ?

File: gefserver/pier/internal/dckr/dckr.go [hunk 4]
```diff
@@ -469,6 +478,9 @@ func (c Client) CreateSwarmService(repoTag string, cmdArgs []string, binds []Vol
 	taken from https://github.com/moby/moby/blob/v1.12.0-rc4/daemon/cluster/executor/container/container.go#L331 */
 	calculatedNanoCPU := (limits.CPUQuota * 1e9) / limits.CPUPeriod
 
+	swarmContext, cancel := context.WithTimeout(context.Background(), time.Duration(executionTimeOut)*time.Second)
+	defer cancel()
```
Comment (emanueldima) [inline gefserver/pier/internal/dckr/dckr.go]: ?

File: gefserver/pier/internal/dckr/dckr.go (hunk 1 above)
Comment (ubuntolog) [inline gefserver/pier/internal/dckr/dckr.go]: https://golang.org/pkg/context/
createContainerContext is a context for creating a container, context with a timeout, defer cancel()  releases resources if slowOperation completes before timeout elapses

File: gefserver/pier/internal/dckr/dckr.go (hunk 2 above)
Comment (ubuntolog) [inline gefserver/pier/internal/dckr/dckr.go]: Right, this one is not needed

File: gefserver/pier/internal/dckr/dckr.go (hunk 3 above)
Comment (ubuntolog) [inline gefserver/pier/internal/dckr/dckr.go]: StartContainer cannot accept a context, StartContainerWithContext is the same thing but with a context

File: gefserver/pier/internal/dckr/dckr.go (hunk 4 above)
Comment (ubuntolog) [inline gefserver/pier/internal/dckr/dckr.go]: Service execution timeout for a swarm service

**Pull request merged**